### ISSUE TITLE
404 Client Error: Not Found ("No such image") error when starting addon

### ISSUE THREAD

**Issue by JJussi** [danielwelch/hassio-zigbee2mqtt#366]
**zigbee2mqtt add-on version (if edge, please report commit hash):**
1.14.2
**Operating environment (HassOS, Virtual Machine, Device/platform running Hass.io):**
VirtualBox, Linux, amd64

**Description of problem:**
There is no amd64 version available

**Your logs from Hass.io**
`20-08-02 10:07:02 ERROR (SyncWorker_0) [supervisor.docker.interface] Can't install dwelch2101/zigbee2mqtt-amd64:1.14.2 -> 404 Client Error: Not Found ("no such image: dwelch2101/zigbee2mqtt-amd64:1.14.2: No such image: dwelch2101/zigbee2mqtt-amd64:1.14.2").`

**Comment by ciotlosm** [comment]
This should be available. It's visible in docker hub: https://hub.docker.com/r/dwelch2101/zigbee2mqtt-amd64/tags 
Can you re-check?

**Comment by ciotlosm** [comment]
There seems to be a more generic problem as see based on the number of issues referencing this error:
https://github.com/danielwelch/hassio-zigbee2mqtt/issues/366 https://github.com/danielwelch/hassio-zigbee2mqtt/issues/365 https://github.com/danielwelch/hassio-zigbee2mqtt/issues/352 https://github.com/danielwelch/hassio-zigbee2mqtt/issues/334 https://github.com/danielwelch/hassio-zigbee2mqtt/issues/293 https://github.com/danielwelch/hassio-zigbee2mqtt/issues/267#issuecomment-558845125 https://github.com/danielwelch/hassio-zigbee2mqtt/issues/340#issuecomment-631652864 https://github.com/danielwelch/hassio-zigbee2mqtt/issues/340#issuecomment-638257269

And potentially also:  https://github.com/danielwelch/hassio-zigbee2mqtt/issues/382

@danielwelch Maybe people are trying to install the version before all builds complete?

**Comment by danielwelch** [comment]
> @danielwelch Maybe people are trying to install the version before all builds complete?
This has almost always been the case (there is an issue like this raised on a lot of releases). May be worth adding something to the issue template to call it out more obviously. Build times used to be a lot longer, and issues like this were big motivators for me to reduce build time.

**Comment by ciotlosm** [comment]
Is there a possible to use build triggers to mark the version available only after all builds are published on docker?

**Comment by ciotlosm** [comment]
Docs recommend doing it manually by using a different branch: https://developers.home-assistant.io/docs/add-ons/publishing#build-scripts-to-publish-add-ons-to-docker-hub

> Home Assistant assumes that the master branch of your add-on repository matches the latest tag on Docker Hub. When you're building a new version, it's suggested that you use another branch, ie build or do it with a PR on GitHub. After you push the add-on to Docker Hub, you can merge this branch to master.

Using a `dev` branch would work if there was no `edge` addon version. One potential solution would be to always use `dev` and on each commit (contributors could use PRs) to trigger a build job that when it finishes can do an automatic PR/Merge to `master`. To solve the `edge` version problem, an automation could be created so on each successful build, a counter is incremented and `edge` has a static numeric value like 11390 which corresponds to the build number. 

User -> commit -> PR to `dev` -> Build is triggered -> push to Docker hub -> Auto-increment `edge` version based on build number & Auto PR to `master`

In the above approach you just have to make sure that `config.json` for stable has the right version and lock `master` so only automatic process can push (no merge errors for being ahead).

**Comment by danielwelch** [comment]
We'd need to limit who could submit a pull request to `dev` or whatever branch we'd use for this, as I don't want to push an image to Docker hub for any pull request.

**Comment by ciotlosm** [comment]
You don't need to. As you would not accept the PR to trigger a build. My workflow above misses the approval stage. You would trigger a build only after the PR is merged in `dev`

**Comment by danielwelch** [comment]
Gotcha, sorry I misunderstood. That sounds like a reasonable approach.

**Comment by danielwelch** [comment]
@ciotlosm another related issue being discussed in #384 (another motivation for a separate branch that I think is more important than what we've been discussing here), I'm not sure what is proposed here would solve. Specifically, we wouldn't want to do an automatic merge to master, as `zigbee2mqtt/config.json` should not be changed on master branch between releases.

**Comment by ciotlosm** [comment]
@danielwelch that was my exact thought after reading through the other thread. I will try to imagine an ironed out process. The only thing that comes to mind now is having version management removed from config.json.

We change `config.json` to look something like this on `dev` branch:
```
{
  "name": "Zigbee2mqtt",
  "version": "[Z2M_VERSION][BUILD_NO]",
  "slug": "zigbee2mqtt",
```

`[Z2M_VERSION]` and `[BUILD_NO]` would be placeholders that we never change.


⇨ User creates PR for `dev` branch
⇨ Automatic check to see if they touched `version` in `config.json` and drop PR if they did
⇨ PR is merged by a person if everything checks out
⇨ Build is triggered 
⇨ Image pushed to Docker hub 
*** Extra process to update both `config.json` for `stable` and `edge` using placeholders***
⇨ Increment edge version to latest build number
⇨ Inject correct zigbee2mqtt version (from Dockerfile) and the build number (or something else)
*** End of extra process ***
⇨ Auto-create & merge PR to master


Right now the only thing I would not be happy with is that the build number would always go up, so stable would look something like:

```
1.14.1.95
1.15.0.96
```

**Comment by danielwelch** [comment]
To clarify, you're suggesting implementing incremented versioning for edge (based on build number) in order to support updates through the UI?

**Comment by ciotlosm** [comment]
@danielwelch that was in my mind, an incremented version for edge, but thinking about it now I'm unsure when `edge` builds get triggered. Nightly? Tied to z2m commits? On every commit on master?

The problem with the solution above is that if config.json requires to be changed on each build for `edge` that would pollute a lot history depending on how often that gets updated, so I have no solution for it for now. 

I will look more in depth if there is any advanced / dev option in Supervisor.

**Comment by ciotlosm** [comment]
Closing this in favour of #399
